drivers/storage/portworx/components.go

```diff
@@ -0,0 +1,1092 @@
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	pxServiceAccountName      = "portworx"
+	pxClusterRoleName         = "portworx"
+	pxClusterRoleBindingName  = "portworx"
+	pxRoleName                = "portworx"
+	pxRoleBindingName         = "portworx"
+	pxServiceName             = "portworx-service"
+	pxAPIServiceName          = "portworx-api"
+	pxAPIDaemonSetName        = "portworx-api"
+	pvcServiceAccountName     = "portworx-pvc-controller"
+	pvcClusterRoleName        = "portworx-pvc-controller"
+	pvcClusterRoleBindingName = "portworx-pvc-controller"
+	pvcDeploymentName         = "portworx-pvc-controller"
+	lhServiceAccountName      = "px-lighthouse"
+	lhClusterRoleName         = "px-lighthouse"
+	lhClusterRoleBindingName  = "px-lighthouse"
+	lhServiceName             = "px-lighthouse"
+	lhDeploymentName          = "px-lighthouse"
+	lhContainerName           = "px-lighthouse"
+)
+
+var (
+	kbVerRegex     = regexp.MustCompile(`^(v\d+\.\d+\.\d+).*`)
+	controllerKind = corev1alpha1.SchemeGroupVersion.WithKind("StorageCluster")
+)
+
+func (p *portworx) installComponents(cluster *corev1alpha1.StorageCluster) error {
+	t, err := newTemplate(cluster)
+	if err != nil {
+		return err
+	}
+
+	if err = p.setupPortworxRBAC(t.cluster); err != nil {
+		return err
+	}
+	if err = p.setupPortworxService(t); err != nil {
+		return err
+	}
+	if err = p.setupPortworxAPI(t); err != nil {
+		return err
+	}
+	if err = p.createCustomResourceDefinitions(); err != nil {
+		return err
+	}
+
+	if t.needsPVCController {
+		if err = p.setupPVCController(t); err != nil {
+			return err
+		}
+	} else {
+		if err = p.removePVCController(t.cluster.Namespace); err != nil {
+			return err
+		}
+	}
+
+	if cluster.Spec.UserInterface != nil && cluster.Spec.UserInterface.Enabled {
+		if err = p.setupLighthouse(t); err != nil {
+			return err
+		}
+	} else {
+		if err = p.removeLighthouse(t.cluster.Namespace); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *portworx) setupPortworxRBAC(cluster *corev1alpha1.StorageCluster) error {
+	ownerRef := metav1.NewControllerRef(cluster, controllerKind)
+	if err := p.createServiceAccount(cluster.Namespace, ownerRef); err != nil {
+		return err
+	}
+	if err := p.createClusterRole(ownerRef); err != nil {
+		return err
+	}
+	if err := p.createClusterRoleBinding(cluster.Namespace, ownerRef); err != nil {
+		return err
+	}
+	if err := p.createRole(cluster.Namespace, ownerRef); err != nil {
+		return err
+	}
+	if err := p.createRoleBinding(cluster.Namespace, ownerRef); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *portworx) setupPortworxService(t *template) error {
+	ownerRef := metav1.NewControllerRef(t.cluster, controllerKind)
+	if !p.portworxServiceCreated {
```

Comment:
For services, deployments and daemonsets the object we create and the api server object are always different because of the defaults. It is difficult to do a diff to check if the object needs to be updated. That is why the CreateOrUpdate only creates the object if not present, else updates.
To avoid updating these objects on every reconcile, I have added the flags.